### PR TITLE
Feature/174 projects

### PR DESCRIPTION
**Pull request by eddiejaoude**
* [x] #208 is merged into `prototype-v0.3` then `prototype-v0.3` onto here, this **PR** will reduce in size

### CODE REVIEW

File: src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java [hunk 1]
```diff
@@ -1,12 +1,71 @@
...
+
+        Project existingProject = findByUid(project.getUid());
+        if (existingProject != null) {
+            if (existingProject.getUser() != user) {
+                throw new PermissionDeniedException("Not owner of Project");
+            }
+
+            existingProject.setName(project.getName());
+            existingProject.setDescription(project.getDescription());
+            existingProject.setIsPrivate(project.getIsPrivate());
```
Comment (eddiejaoude) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: @simoncasey Is there a better way to do the above? Update existing `Project` with *updates*

File: src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java [hunk 2]
```diff
@@ -1,12 +1,71 @@
...
+
+    public Iterable<Project> findAllByPublic() {
+        return this.projectRepository.findAllPublic();
+    }
+
+    public Project findByUid(String uuid) {
+        return this.projectRepository.findByUid(uuid, getCurrentUsername());
+    }
+
+    public Boolean delete(Project project) {
```
Comment (simoncasey) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: why does this need to return anything?

Comment (eddiejaoude) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: Fair point. I will update

File: src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java (hunk 1 above)
Comment (simoncasey) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: This is roughly correct i think

File: src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java [hunk 3]
```diff
@@ -1,12 +1,71 @@
...
 public class ProjectService {
 
     @Autowired
-    ProjectRepository projectRepository;
+    private ProjectRepository projectRepository;
+
+    @Autowired
+    private UserService userService;
+
+    private String getCurrentUsername() {
```
Comment (simoncasey) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: Why is this duplicated within UserService?

File: src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java (hunk 1 above)
Comment (eddiejaoude) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: Ok cool. I was not sure if there was an easier / magical way to do it.

File: src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java (hunk 3 above)
Comment (eddiejaoude) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: I could do it as an `Advice`?

File: src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProjectDefinitions.java [hunk 4]
```diff
@@ -3,10 +3,38 @@
...
 import cucumber.api.java.en.When;
 import io.dashboardhub.pipelinedashboard.PipelinedashboardApplicationTests;
+import io.dashboardhub.pipelinedashboard.domain.Project;
+import io.dashboardhub.pipelinedashboard.domain.User;
+import io.dashboardhub.pipelinedashboard.repository.ProjectRepository;
+import io.dashboardhub.pipelinedashboard.repository.UserRepository;
 import org.junit.Assert;
+import org.springframework.beans.factory.annotation.Autowired;
 
 public class ProjectDefinitions extends PipelinedashboardApplicationTests {
```
Comment (simoncasey) [inline src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProjectDefinitions.java]: I'm confused here. I thought we were using cucumber purely for acceptance tests (full application tests via selenium). Are we now using the same for integration tests (which this class looks like)??

Comment (eddiejaoude) [inline src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProjectDefinitions.java]: This class is seeding the database with dummy data for the Scenario.

No reason why we can't use it for Integration tests too? We do this on other projects.

File: src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java (hunk 3 above)
Comment (simoncasey) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: yup, or just call a method on UserService as you have it in scope at line 18

File: src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProjectDefinitions.java (hunk 4 above)
Comment (simoncasey) [inline src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProjectDefinitions.java]: ok cool

File: src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java (hunk 3 above)
Comment (eddiejaoude) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: :+1:

Comment (eddiejaoude) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: Actually an **Advice** already exists for the **CurrentUser**, it just needs updating with the full User object https://github.com/DashboardHub/PipelineDashboard/blob/prototype-v0.3/src/main/java/io/dashboardhub/pipelinedashboard/controller/advice/CurrentUserAdvice.java

Comment (eddiejaoude) [inline src/main/java/io/dashboardhub/pipelinedashboard/service/ProjectService.java]: I took your advice (no pun intended) and used the **UserService** for now

**Pull request merged**